// member/model/member.js
var _ = require('underscore'),
    logger = require("libs/log")(module),
    I18n = require('i18n-2'),
    mongoose = require('mongoose');

var Schema = mongoose.Schema;
var Member = new Schema({
    name: {
        type: String,
        default: "default name",
        required: true
    },

    email: {
        type: String,
        required: true
    },

    password: {
        type: String,
        required: true
    },

    group: {
        type: String,
        default: "guest",
        required: true
    },

    confirmationId: {
        type: String,
        default: "",
        required: true
    },

    //200 - active
    //403 - blocking
    //400 - not confirmed
    status: {
        type: Number,
        default: 400,
        required: true
    }
});

Member.statics.isHaveConfirmationId = function(confirmationId, cb){
    this.find({confirmationId: confirmationId}, null, function(err, members){
        if( err ){
            logger.error(err);
            cb("Server error");
            return false;
        }

        if( members.length === 0 ){
            cb(null, false);
        }else{
            cb(null, members[0]);
        }
    });
}
Member.statics.isHaveMember = function(email, cb){
    this.find({email: email}, null, function(err, members){
        if( err ){
            logger.error(err);
            cb("Server error");
            return false;
        }

        if( members.length === 0 ){
            cb(null, false);
        }else{
            cb(null, members[0]);
        }
    });
}
Member.statics.comparePassword = function(password, memberPassword){
    return (password == memberPassword) ? true : false;
}
Member.statics.getConfirmationId = function(){
    var random = Math.random() * Math.random() * Math.random();
    return Math.floor((new Date()).getTime() * random) + "";
}
Member.statics.registerNewMember = function(email, password, cb){

    var MemberModel = mongoose.model('Member');
    var confirmationId = this.getConfirmationId();
    var member = new MemberModel({
        email: email,
        password: password,
        confirmationId: confirmationId
    });

    //save user
    member.save(function(err){
        if(err){
            logger.error(err);
            return cb("Server error");
        }
        cb(null, member);
    });
}

Member.methods.confirm = function(cb){
    this.update({
        confirmationId: null,
        status: 200
    }, function(err, member){
        if(err){
            logger.error(err);
            if(cb) cb(err);
        }
        if(cb) cb(null, member);
    })
}

var MemberModel = mongoose.model('Member', Member);
module.exports = MemberModel;
