// member/model/member.js
var _ = require('underscore'),
    logger = require("libs/log")(module),
    EmailSender = require("EmailSender"),
    mongoose = require('mongoose');

var Schema = mongoose.Schema;

var Member = new Schema({
    name: {
        type: String,
        default: "default name",
        required: true
    },

    email: {
        type: String,
        required: true
    },

    password: {
        type: String,
        required: true
    },

    group: {
        type: String,
        default: "guest",
        required: true
    },

    confirmationId: {
        type: String,
        default: "",
        required: true
    },

    //200 - active
    //403 - blocking
    //400 - not confirmed
    status: {
        type: Number,
        default: 400,
        required: true
    }
});

Member.statics.isHaveMember = function(email, cb){
    this.find({email: email}, null, function(err, members){
        if( err ){
            logger.error(err);
            cb("Server error");
            return false;
        }

        if( members.length === 0 ){
            cb(null, false);
        }else{
            cb(null, true);
        }
    });
}
Member.statics.getConfirmationId = function(){
    var random = Math.random() * Math.random() * Math.random();
    return Math.floor((new Date()).getTime() * random) + "";
}

Member.statics.registerNewMember = function(email, password, cb){

    var MemberModel = mongoose.model('Member');
    var confirmationId = this.getConfirmationId();
    var member = new MemberModel({
        email: email,
        password: password,
        confirmationId: confirmationId
    });

    //save user
    member.save(function(err){
        if(err){
            logger.error(err);
            return cb("Server error");
        }
        cb(null);

        //send email with confirmation code
        var emailSender = new EmailSender({
            to: email,
            subject: "Registration",
            text: "Hi! Cool! Go to http://map.batros.in.ua/confirm/" + confirmationId
        });
        emailSender.send(function(err){
            if( err ){
                logger.error(err);
                return false;
            }
        });
    });
}

var MemberModel = mongoose.model('Member', Member);

module.exports = MemberModel;
